### ISSUE TITLE
JSONOutputArchive distructor throws on empty itsNodeStack

### ISSUE THREAD

**Issue by mheyman** [USCiLab/cereal#200]
In json.hpp, the `~JSONOutputArchive()` method does an `itsNodeStack.top()` without verifying that `itsNodeStack.size() > 0`. `itsNodeStack.size()` will equal zero if the `JSONOutputArchive::finishNode()` method gets called on the last node (the `finishNode()` call seems to be required if you wish to get the final '}' on the JSON text prior to destroying the JSONOutputArchive).

**Comment by AzothAmmo** [comment]
What situation are you in where `itsNodeStack.size()` reaches zero when the destructor is called?

**Comment by AzothAmmo** [comment]
I just re-read your issue, you are misusing `finishNode()` and more generally, the archive itself.  As the [documentation states](http://uscilab.github.io/cereal/serialization_archives.html) you should use the archives in an RAII manner - the destruction of the archive will finish any remaining work it has to do.  `finishNode()` is meant for internal use only.

See also: #48, #101, and [this stack overflow post](http://stackoverflow.com/questions/30194212/cereal-json-output-misses-closing-curly-brace) about this same issue.

**Comment by mheyman** [comment]
Thanks for your response.

My final goal was to be able to stream one item after another with the
streamed object getting assembled by passing the archive to a chain of
methods but I was working towards it when I ran into difficulties.

I had code that passed the output archive instance into a method that added
some content to the archive and fired the copy-the-resulting-JSON-elsewhere
event. This, as you know, would send content without the final '}'.

The finishNode() method seemed like it could be used to finish the node
prior to writing. The copy was made and sent off but the later exception
was a disappointment :-( also, I realized that just calling finishNode()
wouldn't allow me to start a new node. I examined cereal and realized that
I could create a:

    JSONOutputArchive::flushAndClear()
    {  // error checking removed
       finishNode();
       itsNodeStack.push(NodeType::StartObject);
    }

This seems to work fine. I can call this after finishing each item and then
continue on with the next root-level item. But, I don't know the cereal
code base well enough to know if I actually would have produced latent bugs
so I didn't think this was a viable solution for our library. So, I removed
the method.

I have now wrapped JSONOutputArchive in my own class that, upon calling
operator(), creates the output archive if it doesn't already exist and
calls the output archive's operator(). It also has an flushAndClear() call
that destroys the wrapped output archive. This is completely cereal
compliant and allows me to stream one item after another but it seems a lot
of wasted effort compared to my earlier solution of adding flushAndClear()
to .JSONOutputArchive.
